stdpopsim/models.py

```diff
@@ -137,6 +137,176 @@ def asdict(self):
+            mig_matrix = N_M["migration_matrix"]
+            C = self._get_C_from_N(pop_sizes)
+            Mp = self._get_Mp_from_M(mig_matrix)
+            G = self._get_G_from_Mp_C(Mp, C)
+            P = P * np.exp(dt * G)
+            r[j] = np.sum(np.matmul(P, C)) / np.sum(P)
+        return steps, r
+
+    def num_pops(self):
+        """
+        This function returns the number populations
+        defined by the demographic model
+        """
+        ddb = msprime.DemographyDebugger(**self.asdict())
+        return len(ddb.epochs[0].populations)
+
+    # TODO Rename
+    def _get_epoch_at_t(self, t):
+        """
+        Given a time, t (in generations), find and return
+        the Epoch for which this belongs.
+        If there is no epoch that exists for the given t
+        then it returns None.
+
+        <class 'int'> -> <class 'msprime.simulations.Epoch'>
+        """
+
+        ddb = msprime.DemographyDebugger(**self.asdict())
+        epochs = ddb.epochs
+        j = 0
+        while epochs[j].end_time <= t:
+            j += 1
+        return j, epochs[j]
+
+    # TODO Rename
+    def _get_N_M_at_t(self, t):
+        """
+        Given a time, t (in generations), find and return
+        a dictionary containing:
+        1: "population_sizes" the vector N which should represent effective
+        population size for each populations at time t and,
+        2: "migration_matrix" The migration matrix for for the populations at time t
+
+        <class 'int'> -> <class 'dict'>
+        """
+
+        epochIndex, epoch = self._get_epoch_at_t(t)
+        ddb = msprime.DemographyDebugger(**self.asdict())
+        N = ddb.population_size_history[:, epochIndex]
+        for i, pop in enumerate(epoch.populations):
+            s = t - epoch.start_time
+            g = pop.growth_rate
+            N[i] *= np.exp(-1 * g * s)
+        return {"population_sizes": N, "migration_matrix": epoch.migration_matrix}
+
+    def _delta(self, x, y):
+        if x == y:
+            return 1
+        else:
+            return 0
+
+    # TODO Rename
+    def _get_C_from_N(self, N):
+        """
+        Compute the matrix C, which reprents coalescent probabilities
+        in each population, from Individual effective population sizes.
+        """
+
+        n = len(N)
+        C = np.zeros([n**2, n**2])
+        index_array = np.array(range(n**2)).reshape([n, n])
+        for idx in range(n):
+            C[index_array[idx, idx], index_array[idx, idx]] = 1 / N[idx]
+        return C
+
+    # TODO Rename
+    def _get_G_from_Mp_C(self, Mp, C):
+        """
+        ????
+        """
+
+        Identity = np.eye(len(Mp[0]))
+        G = (np.kron(Mp, Identity) + np.kron(Identity, Mp)) - C
+        return G
+
+    # TODO Rename
+    def _get_Mp_from_M(self, M):
+        """
+        Simply compute Mp. which represents the **,
+        From the migration matrix.
+        """
+
+        Mp = copy.deepcopy(M)
```

Comment:
What's wrong with
```
Mp = M.copy()
```
?

Reply:
well, because python doesn't pass by reference we don't need to copy at all, I'll fix this.

Reply:
```>>> x = [[1,2],[3,4]]
>>> def yoo_copy(alist):
...  a = alist.copy()
...  a[0][0] = 500
...  return a
... 
>>> yoo_copy(x)
[[500, 2], [3, 4]]
>>> x
[[500, 2], [3, 4]]
>>> x = [[1,2],[3,4]]
>>> def yoo_deepcopy(alist):
...  a = copy.deepcopy(alist)
...  a[0][0] = 500
...  return a
... 
>>> yoo_deepcopy(x)
[[500, 2], [3, 4]]
>>> x
[[1, 2], [3, 4]]```

Reply:
deep copy is reccursive and python is pass by reference with mutable object. Because the migration matrix is a list of lists, it needs to be reccursivly copied

Reply:
ah-ha. thanks! scary that I didn't realize that.